crates/storage/src/lazy/mod.rs

```diff
@@ -148,16 +148,25 @@ where
 
     /// Try to read the `value` from the contract storage.
     ///
+    /// To successfully retrieve the `value`, the encoded `key` and `value`
+    /// must both fit into the static buffer together.
+    ///
     /// Returns:
     /// - `Some(Ok(_))` if `value` was received from storage and could be decoded.
-    /// - `Some(Err(_))` if the encoded length of `value` exceeds the static buffer size.
+    /// - `Some(Err(_))` if retrieving the `value` would exceed the static buffer size.
     /// - `None` if there was no value under this storage key.
     pub fn try_get(&self) -> Option<ink_env::Result<V>> {
-        let encoded_length: usize = ink_env::contains_contract_storage(&KeyType::KEY)?
+        let key_size = KeyType::KEY.encoded_size();
+
+        if key_size >= ink_env::BUFFER_SIZE {
+            return Some(Err(ink_env::Error::BufferTooSmall))
+        }
+
+        let value_size = ink_env::contains_contract_storage(&KeyType::KEY)?
```

Comment:
I think a type annotation `let value_size: usize` is useful here because we are using `try_into`. Even though it is inferred it is good for readability. This is done in `mapping` too https://github.com/paritytech/ink/blob/48ed3f853659704bcf5e3e60aa49cf1acbc0bf4d/crates/storage/src/lazy/mapping.rs#L275